src/hypothesis/strategies.py

```diff
@@ -775,19 +778,126 @@ def seed_random(seed):
+            kwargs[ms] = from_type(hints[ms])
+
     return tuples(tuples(*args), fixed_dictionaries(kwargs)).map(
         lambda value: target(*value[0], **value[1])
     )
 
 
 @cacheable
+def from_type(thing):
+    """Looks up the appropriate search strategy for the given type.
+
+    ``from_type`` is used internally to fill in missing arguments to
+    :func:`~hypothesis.strategies.builds` and can be used interactively
+    to explore what strategies are available or to debug type resolution.
+
+    You can use :func:`~hypothesis.strategies.register_type_strategy` to
+    handle your custom types, or to globally redefine certain strategies -
+    for example excluding NaN from floats, or use timezone-aware instead of
+    naive time and datetime strategies.
+
+    The resolution tries four options, in this order:
+
+    1. If ``thing`` is in the default lookup mapping or user-registered lookup,
+       return the corresponding strategy.  The default lookup covers all types
+       with Hypothesis strategies, including extras where possible.
+    2. If ``thing`` is from the :mod:`python:typing` module, return the
+       corresponding strategy (special logic).
+    3. If ``thing`` has one or more subtypes in the merged lookup, return
+       the union of the strategies for those types that are not subtypes of
+       other elements in the lookup.
+    4. Finally, if ``thing`` has type annotations for all required arguments,
+       it is resolved via :func:`~hypothesis.strategies.builds`.
+
+    """
+    from hypothesis.searchstrategy import types
+    # Look for a known concrete type or user-defined mapping
+    lookup = dict(types.type_strategy_mapping())
+    lookup.update(types.generic_type_strategy_mapping())
+    lookup.update(types._global_type_to_strategy_lookup)
+    if thing in lookup:
+        strategy = lookup[thing]
+        if not isinstance(lookup[thing], SearchStrategy):
+            strategy = lookup[thing](thing)
+        if strategy.is_empty:
+            raise ResolutionFailed(
+                'Error: %r resolved to an empty strategy' % (thing,))
+        return strategy
+    # I tried many more elegant checks, but `typing` tends to treat the type
+    # system as a loose guideline at best so they were all unreliable.
+    if getattr(thing, '__module__', None) == 'typing':
+        return types.from_typing_type(thing)
+    # If there's no exact match above, use similar subtype resolution logic
+    strategies = [  # pragma: no cover
+        v if isinstance(v, SearchStrategy) else v(thing)
+        for k, v in lookup.items()
+        if isclass(k) and isclass(thing) and types.try_issubclass(k, thing) and
+        sum(types.try_issubclass(k, T) for T in lookup) == 1
+    ]
+    empty = ', '.join(repr(s) for s in strategies if s.is_empty)
+    if empty:
+        raise ResolutionFailed(
+            'Could not resolve %s to a strategy; consider using '
+            'register_type_strategy' % empty)
```

Comment:
Since you can't register an empty strategy, it's a guard against failures in `v(thing)` - where something like an unparameterised Union would resolve to `one_of([])`, and hence empty.  An empty strategy seems to me a clear failure, as you can't draw any value from the type, and also provides an elegant way to signal failure.